src/index.js

```diff
@@ -17,76 +19,74 @@ export default function({ types: t }) {
       TaggedTemplateExpression: {
         enter(path, { opts }) {
           const addDisplayName = (opts.displayName === undefined || opts.displayName === null) ? true : opts.displayName
-					const addIdentifier = (opts.ssr === undefined || opts.ssr === null) ? true : opts.ssr
+          const addIdentifier = (opts.ssr === undefined || opts.ssr === null) ? true : opts.ssr
           const tag = path.node.tag
 
           if (!isStyled(tag)) return
-          if (path.node._styledComponentsSeen) {
+          if (path._styledComponentsSeen) {
             return
           }
 
-          let displayName
-
-          if (addDisplayName) {
-            path.find((path) => {
-              if (path.isAssignmentExpression()) {
-                displayName = path.node.left
-              } else if (path.isObjectProperty()) {
-                displayName = path.node.key
-              } else if (path.isVariableDeclarator()) {
-                displayName = path.node.id
-              } else if (path.isStatement()) {
-                // we've hit a statement, we should stop crawling up
-                return true
-              }
+          let displayName, target
 
-              // we've got an displayName! no need to continue
-              if (displayName) return true
-            })
+          path.find((path) => {
+            // const X = styled
+            if (path.isAssignmentExpression()) {
+              target = getTarget(path.node.right.tag)
+              displayName = path.node.left
+            // const X = { Y: styled }
+            } else if (path.isObjectProperty()) {
+              target = getTarget(path.node.value.tag)
+              displayName = path.node.key
+            // let X; X = styled
+            } else if (path.isVariableDeclarator()) {
+              target = getTarget(path.node.init.tag)
+              displayName = path.node.id
+            } else if (path.isStatement()) {
+              // we've hit a statement, we should stop crawling up
+              return true
+            }
 
-            // ensure that we have an displayName we can inherit from
-            if (!displayName) return
+            // we've got an displayName (if we need it) and a target! no need to continue
+            if ((!addDisplayName || displayName) && target) return true
+          })
 
-            // foo.bar -> bar
-            if (t.isMemberExpression(displayName)) {
-              displayName = displayName.property
-            }
+          // foo.bar -> bar
+          if (t.isMemberExpression(displayName)) {
+            displayName = displayName.property
+          }
 
-            // identifiers are the only thing we can reliably get a name from
-            if (!t.isIdentifier(displayName)) {
-              displayName = undefined
-            } else {
-              displayName = displayName.name
-            }
+          // styled call without variable
+          if (!target) {
+            target = getTarget(path.node.tag)
```

Comment:
getting target from path.node.tag works for all tests. In which cases you need to determine it differently?